### PR TITLE
Fix clearing cache after receiving channel truncated event

### PR DESCRIPTION
**Pull request by adasiewiczr**
### 🎯 Goal

Fix clearing cache after receiving channel truncated event.

### 🛠 Implementation details

1. Replace `mutableState.messageList` with `mutableState._messages` when parsing messages to prevent propagating wrong state when updating `mutable._messages`
2. Remove `channel` from LRU cache when it was truncated
3. Fix updating `lastSyncAt` when recovery was run but there were no active channels

### 🎨 UI Changes

<table>
<thead>
<tr>
<th>Before</th>
<th>After</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<video src="https://user-images.githubusercontent.com/17440581/151767617-20505ab2-efda-4521-b2ff-b39b4b2eea79.mp4" controls="controls" muted="muted" />
</td>
<td>
<video src="https://user-images.githubusercontent.com/17440581/151767600-3564f2a9-1ef0-4d20-8877-71f936dd8702.mp4" controls="controls" muted="muted" />
</td>
</tr>
</tbody>
</table>

### 🧪 Testing

You will need two devices and apply a patch on one of them (it adds truncating channel after clicking on user avatar in `ChatFragment`).

1. Turn off `OfflinePlugin` - use `ChatDomain` instead
2. Send a message from device A and turn on airplane mode
2. Click on user avatar on device B (truncate channel)
3. Turn off airplane mode on device A and observe channel is properly truncated after the sync
4. Go back to channel list and click on the same channel one more time -> observe channel is properly truncated

<details>
  <summary>Patch file to apply these changes for testing</summary>

```
Index: stream-chat-android-ui-components-sample/src/main/kotlin/io/getstream/chat/ui/sample/feature/chat/ChatViewModel.kt
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/stream-chat-android-ui-components-sample/src/main/kotlin/io/getstream/chat/ui/sample/feature/chat/ChatViewModel.kt b/stream-chat-android-ui-components-sample/src/main/kotlin/io/getstream/chat/ui/sample/feature/chat/ChatViewModel.kt
--- a/stream-chat-android-ui-components-sample/src/main/kotlin/io/getstream/chat/ui/sample/feature/chat/ChatViewModel.kt	(revision 851d0f068967850baf7f5e32a5902563011d122f)
+++ b/stream-chat-android-ui-components-sample/src/main/kotlin/io/getstream/chat/ui/sample/feature/chat/ChatViewModel.kt	(date 1642542167472)
@@ -4,11 +4,12 @@
 import androidx.lifecycle.MediatorLiveData
 import androidx.lifecycle.MutableLiveData
 import androidx.lifecycle.ViewModel
+import io.getstream.chat.android.client.ChatClient
 import io.getstream.chat.android.client.models.Member
+import io.getstream.chat.android.client.models.Message
 import io.getstream.chat.android.livedata.ChatDomain
 import io.getstream.chat.android.livedata.controller.ChannelController
 import io.getstream.chat.android.livedata.utils.Event
-import io.getstream.chat.ui.sample.util.extensions.isAnonymousChannel
 
 class ChatViewModel(private val cid: String, private val chatDomain: ChatDomain = ChatDomain.instance()) : ViewModel() {
 
@@ -31,16 +32,10 @@
     fun onAction(action: Action) {
         when (action) {
             is Action.HeaderClicked -> {
-                val controller = requireNotNull(channelController)
-                _navigationEvent.value = Event(
-                    if (action.members.size > 2 ||
-                        controller.offlineChannelData.value?.isAnonymousChannel() == false
-                    ) {
-                        NavigationEvent.NavigateToGroupChatInfo(cid)
-                    } else {
-                        NavigationEvent.NavigateToChatInfo(cid)
-                    }
-                )
+                ChatClient.instance()
+                    .channel(cid)
+                    .truncate(Message(text = "System message"))
+                    .enqueue()
             }
         }
     }

```

</details>

#### General
- [x] I have signed the [Stream CLA](https://docs.google.com/forms/d/e/1FAIpQLScFKsKkAJI7mhCr7K9rEIOpqIDThrWxuvxnwUq2XkHyG154vQ/viewform) (required)
- [x] Assigned a person / code owner group (required)
- [x] Thread with the PR link started in a respective Slack channel (#android-chat-core or #android-chat-ui) (required)
- [ ] ~PR targets the `develop` branch~
- [ ] ~PR is linked to the GitHub issue it resolves~

#### Code & documentation
- [x] Changelog is updated with client-facing changes
- [ ] ~New code is covered by unit tests~
- [x] Comparison screenshots added for visual changes
- [ ] ~Affected documentation updated (KDocs, docusaurus, tutorial)~

### ☑️Reviewer Checklist
- [ ] UI Components sample runs & works
- [ ] Compose sample runs & works
- [ ] Bugs validated (bugfixes)
- [ ] Release notes and docs clearly describe changes
- [ ] All code we touched has new or updated KDocs

### CODE REVIEW

File: stream-chat-android-offline/src/main/java/io/getstream/chat/android/offline/ChatDomainImpl.kt [hunk 1]
```diff
@@ -574,14 +574,18 @@ internal class ChatDomainImpl internal constructor(
...
-                    syncStateFlow.value?.let { syncStateFlow.value = it.copy(lastSyncedAt = now) }
+                    updateLastSyncDate(now)
                 }
             }
         } else {
             Result(emptyList())
         }
     }
 
+    private fun updateLastSyncDate(date: Date) {
```
Comment (filbabic) [inline stream-chat-android-offline/src/main/java/io/getstream/chat/android/offline/ChatDomainImpl.kt:585]: Small doc here could be nice

File: stream-chat-android-offline/src/main/java/io/getstream/chat/android/offline/ChatDomainImpl.kt [hunk 2]
```diff
@@ -574,14 +574,18 @@ internal class ChatDomainImpl internal constructor(
...
+        syncStateFlow.value?.let { syncStateFlow.value = it.copy(lastSyncedAt = date) }
```
Comment (JcMinarro) [inline stream-chat-android-offline/src/main/java/io/getstream/chat/android/offline/ChatDomainImpl.kt:586]: Shouldn't we update it with some "backend date"? :thinking: 
Maybe we can open task to review it later

File: stream-chat-android-offline/src/main/java/io/getstream/chat/android/offline/experimental/channel/logic/ChannelLogic.kt [hunk 3]
```diff
@@ -256,7 +256,7 @@ internal class ChannelLogic(
     }
 
     private fun parseMessages(messages: List<Message>): Map<String, Message> {
```
Comment (filbabic) [inline stream-chat-android-offline/src/main/java/io/getstream/chat/android/offline/experimental/channel/logic/ChannelLogic.kt:258]: Same here - I would also love to understand the change between the message list and the _messages here

File: stream-chat-android-offline/src/main/java/io/getstream/chat/android/offline/event/EventHandlerImpl.kt [hunk 4]
```diff
@@ -389,9 +389,11 @@ internal class EventHandlerImpl(
             when (event) {
                 is NotificationChannelTruncatedEvent -> {
                     domainImpl.repos.deleteChannelMessagesBefore(event.cid, event.createdAt)
+                    domainImpl.repos.evictChannel(event.cid)
                 }
                 is ChannelTruncatedEvent -> {
                     domainImpl.repos.deleteChannelMessagesBefore(event.cid, event.createdAt)
+                    domainImpl.repos.evictChannel(event.cid)
```
Comment (JcMinarro) [inline stream-chat-android-offline/src/main/java/io/getstream/chat/android/offline/event/EventHandlerImpl.kt:396]: Can we move this logic into `RepositoryFacade??
Whenever we "delete channels messages before a given date", they should be removed from the local cache as well. If we keep it as two different methods, every time we want to perform the first action, we need to remember that the second action need to be done as well.

**Review by JcMinarro — CHANGES_REQUESTED** [review]

File: stream-chat-android-offline/src/main/java/io/getstream/chat/android/offline/ChatDomainImpl.kt (hunk 2 above)
Comment (adasiewiczr) [reply stream-chat-android-offline/src/main/java/io/getstream/chat/android/offline/ChatDomainImpl.kt:586]: Agree, we should take a look on that in a separate task

**Comment by filbabic** [comment]
Just to leave this here.

One question - I imagine the ChatDomain/OfflinePlugin thing is because we don't process all events properly based on which one is enabled. As sometimes I see that the event was received on Compose but the channel isn't empty. 

### ☑️Reviewer Checklist
- [x] UI Components sample runs & works
- [ ] Compose sample runs & works
- [x] Bugs validated (bugfixes)
- [x] Release notes and docs clearly describe changes
- [ ] All code we touched has new or updated KDocs

**Comment by adasiewiczr** [comment]
@filbabic proper events handling in OfflinePlugin wasn't merged to main yet

File: stream-chat-android-offline/src/main/java/io/getstream/chat/android/offline/experimental/channel/logic/ChannelLogic.kt [hunk 5]
```diff
@@ -255,8 +255,14 @@ internal class ChannelLogic(
         }
     }
 
+    /**
+     * Updates [ChannelMutableState._messages] with new messages.
+     * The message will by only updated if it's creation/update date is newer than the one stored in the StateFlow.
```
Comment (filbabic) [inline stream-chat-android-offline/src/main/java/io/getstream/chat/android/offline/experimental/channel/logic/ChannelLogic.kt:260]:
```suggestion
     * The message will by only updated if its creation/update date is newer than the one stored in the StateFlow.
```

**Review by JcMinarro — APPROVED** [review]

**Review by filbabic — APPROVED** [review]
Looks good - left one small typo suggestion!

**Pull request merged**